src/main/java/org/sonarlint/intellij/editor/SonarExternalAnnotator.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * SonarLint for IntelliJ IDEA
+ * Copyright (C) 2015 SonarSource
+ * sonarlint@sonarsource.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonarlint.intellij.editor;
+
+import com.intellij.codeInsight.hint.InspectionDescriptionLinkHandler;
+import com.intellij.lang.annotation.Annotation;
+import com.intellij.lang.annotation.AnnotationHolder;
+import com.intellij.lang.annotation.ExternalAnnotator;
+import com.intellij.lang.annotation.HighlightSeverity;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.editor.RangeMarker;
+import com.intellij.openapi.editor.markup.EffectType;
+import com.intellij.openapi.editor.markup.TextAttributes;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.ui.JBColor;
+import com.intellij.util.ui.UIUtil;
+import com.intellij.xml.util.XmlStringUtil;
+import org.jetbrains.annotations.NonNls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.sonar.runner.api.Issue;
+import org.sonarlint.intellij.issue.IssueStore;
+
+import java.awt.*;
+import java.util.Collection;
+
+/**
+ * When requested, picks up issues from the {@link IssueStore} for a file.
+ * Note that external annotators are <b>not</b> called if there are errors reported by annotators, which run before (such as compilation errors).
+ *
+ */
+public class SonarExternalAnnotator extends ExternalAnnotator<SonarExternalAnnotator.AnnotationContext, SonarExternalAnnotator.AnnotationContext> {
+  private final boolean unitTest;
+
+  public SonarExternalAnnotator() {
+    this(false);
+  }
+
+  public SonarExternalAnnotator(boolean unitTest) {
+    this.unitTest = unitTest;
+  }
+
+  @Override
+  public void apply(@NotNull PsiFile file, AnnotationContext annotationResult, @NotNull AnnotationHolder holder) {
+    Collection<IssueStore.StoredIssue> issues = annotationResult.store.getForFile(file);
+    for (IssueStore.StoredIssue i : issues) {
+      // reject non-null ranges that are no longer valid. It probably means that they were deleted from the file.
+      if (i.range() == null || i.range().isValid()) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'range' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)